### PR TITLE
Issue #349: support easy debugging of tern.js server from within running Eclipse.

### PR DESCRIPTION
**Pull request by piotrtomiak**
This PR introduces functionality to debug tern.js server located in the workspace with a debugger of choice. So far a connection with Webclipse debugger is set up. A known issue is that Webclipse 0.9.3 will use it's default configured Node.js installation instead of the one specified on Tern Node.js page. Webclipse 0.9.4+ will not have that issue. Screenshots (taken in MyEclipse):

tern-pref-page

tern-locate-file

tern-debug

### CODE REVIEW

File: eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import tern.eclipse.ide.server.nodejs.internal.core.debugging.webclipse.WebclipseDebugger;
+
+public class NodejsDebuggersManager {
+
+	private static Map<String, INodejsDebugger> debuggers = new LinkedHashMap<String, INodejsDebugger>();
+
+	static {
+		// Add supported debuggers
+		addDebugger(new WebclipseDebugger());
```
Comment (angelozerr) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: Instead havinh hard coded list of debugger (and it could be not available), I think an extension point should be more better. What do you think about that?

Comment (piotrtomiak) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: I wanted to avoid any dependencies on debugger plugins, so that they do not have to be modified in any way, or even depend on tern.java during compilation time. I don't like hardcoded things, but in this case an extension point consumed solely by the plugin providing it would be an overkill. However, if there will be a need for other debuggers to use an extension point, and depend on Tern Node.js plugin, one who develops it, can always enhance NodejsDebuggersManager.

Comment (piotrtomiak) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: Oh, and if debugger is not installed it is appropriately shown in the UI (with <not installed> suffix). That way user can see, which JS Debuggers are supported and get one of them.

Comment (angelozerr) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: My big problem is that Webclipse will be available in the combo although users could not have this installed debugger. IMHO I think an extension point should be better. I understand your problem that you don't want to have dependenies to tern.java inside Webclipse, so here my suggestion : 

 * create an extension point for tern debugger
 * create a new folder https://github.com/angelozerr/tern.java/tree/master/eclipse/debugger
 * create a new project tern.eclipse.ide.debugger.webclipse which will host WebclipseDebugger

We could have too a new project  tern.eclipse.ide.debugger.core which will provide some abstract classes with launch, but for the moment we have one debugger.

After that we will able to create a new features debugger, and user will able to choose the debugger to install.

What do you think about that?

Comment (piotrtomiak) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: > My big problem is that Webclipse will be available in the combo although users could not have this installed debugger.

Ok, so my idea is the opposite. I think that it's good for a user to see what options he has, otherwise he will not understand why he is not able to debug, or worse, he will not know which debugger to get. Note, I've update commit to have a validation logic for the field.

tern-debug-not-installed

Also, if there is interest from other debugger providers and they feel they need more flexibility, they can expand the framework. At this point it's hard to say what others would need. I usually refactor the code when it's needed, otherwise I can put a lot of unneeded effort.

Comment (angelozerr) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: At first, thanks @piotrtomiak for your work. We have each an opinion : pros/cons extension point and before merging your code, I would like just to know if other guys like JBoss guys are OK with your proposition to have debugger list hard coded (with uninstalled label of debugger if it is not installed).

@vrubezhny @dgolovin @maxandersen @pascalleclercq @gamerson @PaulVI @fbricon @piotrtomiak @jabby are you OK to have default hard coded debugger list (like Webclipse) or do you prefer having an extension point to fill debugger combo (it means that we must create a new plugin which uses the extension point to provide a debugger)?

Comment (piotrtomiak) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: To be clear, I am not against doing the extension point and setting up plugins. It's just that it's quite a lot of work to do it, I don't see any immediate gains and I've already spend some time to get this done. Current structure of code allows for easy refactoring to have an ext. point. It means that if someone needs ext. point, he can contribute.

However, it is important for us to have Webclipse displayed, even if it is not installed. That way people can know that there is something like Webclipse and that it supports debugging of Tern.js.

Comment (angelozerr) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: >  I don't see any immediate gains and I've already spend some time to get this done.

Yes @piotrtomiak I know. I would like just to know the opinion of other guys. If they prefer having an extension point, I will do that, no problem.

> However, it is important for us to have Webclipse displayed, even if it is not installed. 

Yes I understand. My idea is to keep your code (with Uninstalled, etc) and move to a dedicated plugin. In your case, you will just add this plugin to MyEclipse to have every time Webclipse debugger.

To be honnest with you, I don't know if JBoss guys will be OK to have Webclipse debugger inside JBoss Studio, it's just my fear. 

In other words I don't want to have trouble with JBoss guys and you. I want to find the best solution for JBoss and MyEclipse. Hope you will understand and sorry to make you waste your time.

Comment (piotrtomiak) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: > In other words I don't want to have trouble with JBoss guys and you. I want to find the best solution for JBoss and MyEclipse.

Sure :) Just to be clear, Webclipse (https://marketplace.eclipse.org/content/webclipse) is a different product (it's not MyEclipse, though MyEclipse consumes it). It works with any Eclipse with WTP, including JBoss IDE and right now it's free (one needs to sign up as early adopter). The JS Debugger is based on JSDT, so there should be no conflicts.

Comment (maxandersen) [inline eclipse/tern.eclipse.ide.server.nodejs.core/src/tern/eclipse/ide/server/nodejs/core/NodejsDebuggersManager.java]: If it only says "not installed" and thus the entry in the combo is useless without actually installing it I see this more as noise in the UI.

The extension point is how all other situations like this tend to be resolved.

I agree it would be nice to have visibility into what features you might possibly install in various parts of eclipse - but listing them all and in different ways is not helping the usability nor ecosystem IMO.

+1 for using an extension point instead.

**Comment by angelozerr** [comment]
> +1 for using an extension point instead.

@piotrtomiak I think we should use Extension Point. I will accept your PR and move your WebclipseDebugger inside a plugin. After that you will able to install it inside MyEclipse. Is it OK for you?

I will see that perhaps next week.

**Comment by piotrtomiak** [comment]
> After that you will able to install it inside MyEclipse.

I am already using it for my development, so this is not something I need. Extension point sounds good. However, for me it looks like the Webclipse debugger code could alway be shipped with tern.java, so I don't see any need to extract the code to other plugin and still have it wired through an ext point. So, if we can get away without a new plugin, I can do the ext point stuff, otherwise I really don't want to mess with project setup etc.

As for Webclipse being exposed in UI. What's the point of adding a feature and not giving users any hint on how to use it? If not a combo box item saying not installed (which is not that bad on a page with exposure to really advanced users), than maybe some other way, like a link below combo to a list of supported debuggers (on github wiki, e.g.). This might be actually much more useful, as on github wiki one could add links to the marketplace and some help. What do you think?

**Comment by angelozerr** [comment]
@piotrtomiak in short answer, I don't know if JBoss Tools will be OK to have a combo where debugger like Webclipse are not hosted by JBoss Tools. The extension point in a dedicated plugin gives the capability to a software like JBoss Tools, MyEclipse to choose which debuggers they want to display in the combo. Do you understand what I mean? 

My idea is that tern.java provides all debuggers (and shows it in the update site) and after each software choose the debugger they wish to host in their software by hosting the well debuger plugin.

**Comment by piotrtomiak** [comment]
@angelozerr your project, your call. Next time, I'll just make sure we agree on how it will look like, before I start a work. Update site exposure is at least something.

**Comment by fbricon** [comment]
+100 for an extension point.

What I suggest is setting up a discovery catalog of extensions, listing
available debuggers and matching IUs.
You can populate the combo by fetching/reading the catalog (should be
relatively fast). If IUs are installed nothing to do, else display/enable
an install button.

The catalog url could be overridden by products, either to limit or to
expand choices.

It's more work, but should make everyone happy, eventually.


On Mon, Oct 12, 2015 at 11:49 AM, Angelo <notifications@github.com> wrote:

> @piotrtomiak <https://github.com/piotrtomiak> in short answer, I don't
> know if JBoss Tools will be OK to have a combo where debugger like
> Webclipse are not hosted by JBoss Tools. The extension point in a dedicated
> plugin gives the capability to a software like JBoss Tools, MyEclipse to
> choose which debuggers they want to display in the combo. Do you understand
> what I mean?
>
> My idea is that tern.java provides all debuggers (and shows it in the
> update site) and after each software choose the debugger they wish to host
> in their software by hosting the well debuger plugin.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angelozerr/tern.java/pull/351#issuecomment-147443806>.
>



-- 
"Have you tried turning it off and on again" - The IT Crowd
And if that fails, then http://goo.gl/tnBgH5

**Comment by angelozerr** [comment]
> @angelozerr your project, your call. Next time, I'll just make sure we agree on how it will look like, before I start a work. Update site exposure is at least something.

I'm really sorry @piotrtomiak  -(

> +100 for an extension point.

Is there any chance to have a PR? I would like to accept this PR, but it means that there will have Webclipse inside JBoss Tools. Are you @vrubezhny @dgolovin @maxandersen @mickaelistria  @fbricon OK with that, and after you (or me) develop the extension point?

**Comment by mickaelistria** [comment]
If there is an extension point, and if Webclipse integration is contributed in its own plugin, which would not be included in main Tern.java features, then embedders (such as JBoss Tools) will be able to pick only the part they want and to skip the Webclipse integration if they don't want it.
So the extension point and specific plugin for Webclipse are really required for our ability to keep on using tern.java in JBoss Tools and JBoss Developer Studio, and also to maybe move tern.java to Eclipse Foundation one day.
However, I have no objection to showing the Webclipse integration in the tern.java p2 repository, and I believe that showing all possible integrations (commercial and OSS) as part of tern.java p2 repository is a good thing for highlighting its extensibility and business-friendliness.

The discovery catalog mentioned by @fbricon would be nice, but it's not a requirement for the feature proposed by @piotrtomiak . So let's just keep this topic out for discussing this PR and one can open a new issue dedicated to that topic.

I believe since @piotrtomiak implemented the feature, he's be the best one to design the extension point and interfaces so that they'd at least work for the Webclipse integration. @piotrtomiak Could you take care of it as part of this same PR ?

**Comment by maxandersen** [comment]
I dont understand how there can ever be webclipse in jboss tools or tern.java since webclipse is not free to be bundled, even referenced from tern.java. Unless webclipse has no trademark restrictions and are under an OSS license ?

As I understand it myeclipse is not releasing it for libre, right ?

**Comment by angelozerr** [comment]
> @piotrtomiak Could you take care of it as part of this same PR ?

@mickaelistria as @piotrtomiak told us, he has passed a long time to develop this PR and I'm not sure that he has time to update this PR with extension point. Are you OK I accept this PR (very exciting to see debug features with tern) and when I will have time I will delegate WebclipseDebugger in an extension point?

**Comment by dgolovin** [comment]
IMO extension point is the best option, then we can add discovery catalog with extensions. Fred's suggestion looks good to me.

**Comment by mickaelistria** [comment]
@angelozerr Keep in mind that we'll probably be not able to consume newer builds of tern.java in JBoss Tools and JBoss Developer Studio while they show an advertisement for a commercial product. It's a blocker for adopters.
All due respect to @piotrtomiak work, it's how things are in the Eclipse ecosystem, and in many OSS ecosystems: if one wants some content to go upstream, it must be designed for vendor-neutrality and extensibility. I'm not the project lead here, but in my opinion, not willing to turn code extensible and vendor-neutral would be a blocker for merging a contribution. I let @piotrtomiak consider how much time he has saved by adopting good extensible code written by others compared to how much time it would cost to make his own contribution extensible too. I believe the excuse of "it already took much a lot of time" is not a very appropriate one.

**Comment by angelozerr** [comment]
> It's a blocker for adopters.

Ok, I would like to provide this great debugger feature in 1.1.0. Here my proposition:

 * @piotrtomiak have time to provide an extension point in this PR.
 * @piotrtomiak have NO time to do that, so I would like to accept this PR and comment the code which add WebclipseDebugger with hard code. And when I will have time, I will develop this extension point.

 * @mickaelistria are you OK?
 * @piotrtomiak could you tell me if you will able to have time to develop this extension point, otherwise I will do like I have suggested (accept your PR + comment the add of WebclipseDebugger).

Thanks for your answer.

**Comment by piotrtomiak** [comment]
@angelozerr I'll make the required changes. I'll try to find time this week and provide updated PR ASAP. I might need your help with the new plugins setup though.

@mickaelistria sorry, I didn't think that putting Webclipse in a rather rarely used property page would be a problem for JBoss. I'll keep that in mind. I just thought that would be good for tern plugin developers to know about a good debugger for Node.js. And of course, I do agree that the extension point is the best solution, as I stated earlier.

**Comment by angelozerr** [comment]
< @angelozerr I'll make the required changes. I'll try to find time this week and provide updated PR ASAP. 

Cooool!!! It's a really great news!

> I might need your help with the new plugins setup though.

Yes sure! I will try to help you, no problem. it should be cool if you could create https://github.com/angelozerr/tern.java/tree/master/debugger and set your Webclipse debugger plugin at tern.eclipse.ide.debugger.webclipse 

We could set all debuggers inside debuger folder. For update site, it should be cool to have a debugger folder to.

@paulvi do you think it will possible for you to implement a debugger with nodeclipse once @piotrtomiak will provide the extension point?

**Comment by piotrtomiak** [comment]
> @piotrtomiak will provide the extension point?

Yes, it will be through an extension point.

**Comment by piotrtomiak** [comment]
@angelozerr when do you plan 1.1.0 release? I am unfortunately fully occupied for at least two more days...

**Comment by angelozerr** [comment]
> I am unfortunately fully occupied for at least two more days...

I think I will release 1.1.0 on start of next month, so I will wait for your PR, otherwise I will create the release without your PR.

**Comment by piotrtomiak** [comment]
I've created the plugin and extension point. Creating feature and setting up update site, I am leaving to you :)

**Comment by piotrtomiak** [comment]
@paulvi I've actually added support for Nodeclipse debugger as well. However, please add support for arbitrary nodeclipse path provided by launch configuration of id "nodeinstall_path" to debu launch delegate.

**Comment by mickaelistria** [comment]
I had a look at the "layout" of your contribution (didn't get into details, I was just trying to get how things are structured), and I believe it's very good. Implementations are well separated to interfaces and UI, making it easily extensible. Perfect!
I also like the fact that I've seen many references to the pure Eclipse Debug framework. I believe it's a very good thing to reuse those parts.
Thanks @piotrtomiak !

Your contributions has inspired some food for thought that we had the opportunity to discuss when we met with @angelozerr and @maxandersen at the EclipseDay Grenoble: it would be interesting to start the nodejs process that's hosting Tern as a JavaScript Run/Debug Configuration (provided by JSDT). That way, instead of creating some Debug Configuration specific to Tern, we could imagine implementing plain JS debug configuration relying on JSDT Run/Debug that could be used both for Tern and for regular JavaScript debug.
That's only an idea/suggestion, not something to consider as part of this change.

**Comment by angelozerr** [comment]
Wow, many thanks @piotrtomiak! I trust you so I have accepted the PR. I will play with yoru work and give your feedback.

For your information, I'm studying how to we could use tsserver (TypeScript Server) with tern https://github.com/angelozerr/tern-tsserver to try to support Angular2

Perhaps we should support both node.js process with :

 * http : like today (Tern works liek this)
 * readline (use stout of the process) (TypeScript tsserver works like this).

> hat we had the opportunity to discuss when we met with @angelozerr and @maxandersen at the EclipseDay Grenoble

Yes it was a greag meeting:) 

> it would be interesting to start the nodejs process that's hosting Tern as a JavaScript Run/Debug Configuration (provided by JSDT). 

Yes I will experiment that when I will find time. Today I'm very busy with my study with TypeScript.

Thanks @piotrtomiak again for your great work!

**Comment by piotrtomiak** [comment]
> For your information, I'm studying how to we could use tsserver (TypeScript Server) with tern https://github.com/angelozerr/tern-tsserver to try to support Angular2

Interesting. We will definitely be looking into Angular 2 support.

> Wow, many thanks @piotrtomiak! I trust you so I have accepted the PR. I will play with yoru work and give your feedback.

I want to add images to https://github.com/angelozerr/tern.java/wiki/Debugging-Tern.js-on-Node.js, so that the process of setting up debugger is easier to understand. How do you add images to wiki pages? Once that page is finished, it might be worth removing https://github.com/angelozerr/tern.java/wiki/Debugging-tern.js-with-Nodeclipse .

> Today I'm very busy with my study with TypeScript.

Please remember about creating features and configuring update site. I prefer to leave this sort of stuff to you as the owner of project :)

> That way, instead of creating some Debug Configuration specific to Tern, we could imagine implementing plain JS debug configuration relying on JSDT Run/Debug that could be used both for Tern and for regular JavaScript debug.

To be precise, there is no specific launch configuration type for Tern. It is using the same launch configuration as the debugger itself (be that Nodeclipse or Webclipse). To use plain JSDT launch configuration for node.js debugging, one would have to implement node.js protocol in JSDT debug model, or use chromium SDK one and perform translation to JSDT debug model. In either case that's a lot of work :)

**Comment by angelozerr** [comment]
> Interesting. We will definitely be looking into Angular 2 support.

IMHO, I think we should support that, otherwise Eclipse IDE will be in late again compare to Visual Studio or WebStorm.

> How do you add images to wiki pages? 

I'm not sure that Github allows that? The wiki is on git at https://github.com/angelozerr/tern.java.wiki (images are stored inside images folder). If you cannot do that, send me your images, I will add it.

> Please remember about creating features and configuring update site. I prefer to leave this sort of stuff to you as the owner of project :)

Yes I had already started that, see https://github.com/angelozerr/tern.java/commit/02457bd274117f1a08c85017b2e0a803cb9f845e

I have seen that you have implemented debugger with Nodeclipse, great!

I have added the debugger feature "Tern Debuggers" in the "Tern Tooling" category. 

terndebuggers

@mickaelistria @maxandersen it means that Webclipse and Nodeclipse debugger will be installed if user check "Tern Debuggers". Is it OK?

I have done that, because I would like to avoid adding a new categories "Tern Debuggers" in the update site.

**Comment by piotrtomiak** [comment]
> I'm not sure that Github allows that? The wiki is on git at https://github.com/angelozerr/tern.java.wiki (images are stored inside images folder). If you cannot do that, send me your images, I will add it.

Thanks for info. I don't have access to the repository though. I will send images to you once I get to work on it, probably tomorrow.

**Comment by angelozerr** [comment]
Sorry, the url is https://github.com/angelozerr/tern.java.wiki.git

**Comment by fbricon** [comment]
You can add remote images in markdown pages like :
Tooltip

On Tue, Oct 27, 2015 at 11:13 AM, Angelo <notifications@github.com> wrote:

> Sorry, the url is https://github.com/angelozerr/tern.java.wiki.git
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angelozerr/tern.java/pull/351#issuecomment-151536298>.
>



-- 
"Have you tried turning it off and on again" - The IT Crowd
And if that fails, then http://goo.gl/tnBgH5

**Comment by angelozerr** [comment]
@fbricon thanks for the suggestion, but I will prefer hosting images in the tern.java wiki.
